### PR TITLE
diff: generate check task from dm

### PR DESCRIPTION
**Pull request by WangXiangUSTC**
### What problem does this PR solve? 

support check table by dm's task config https://github.com/pingcap/dm/issues/767

### What is changed and how it works?

get dm task config from dm's HTTP API, and then generate check task for it.

### Check List 

Tests 

 - Unit test
 - Manual test (start a dm task, and use diff check data)

### CODE REVIEW

**Comment by WangXiangUSTC** [comment]
@lance6716 @csuzhangxc PTAL

File: sync_diff_inspector/diff.go [hunk 1]
```diff
@@ -100,6 +113,13 @@ func (df *Diff) init(cfg *Config) (err error) {
 
 // CreateDBConn creates db connections for source and target.
 func (df *Diff) CreateDBConn(cfg *Config) (err error) {
+	if df.subTaskCfgs != nil && len(df.subTaskCfgs) != 0 {
```
Comment (lance6716) [inline sync_diff_inspector/diff.go]: `len()` could handle `len(nil)`, so `df.subTaskCfgs != nil` is redundant.

File: sync_diff_inspector/diff.go [hunk 2]
```diff
@@ -123,13 +143,125 @@ func (df *Diff) CreateDBConn(cfg *Config) (err error) {
...
+					Schema:     schema,
+					Table:      table,
+				})
+			}
+		}
+	}
+
+	for schema, tables := range sourceTablesMap {
+		for table, sourceTables := range tables {
+			if sourceTables == nil || len(sourceTables) == 0 {
```
Comment (lance6716) [inline sync_diff_inspector/diff.go]: ditto

File: sync_diff_inspector/diff.go [hunk 3]
```diff
@@ -123,13 +143,125 @@ func (df *Diff) CreateDBConn(cfg *Config) (err error) {
...
+				continue
+			}
+
+			// TODO: support sql-mode
```
Comment (lance6716) [inline sync_diff_inspector/diff.go]: seems there's only `EnableANSIQuotes` in `SubTaskConfig`

File: sync_diff_inspector/dm_test.go [hunk 4]
```diff
@@ -0,0 +1,165 @@
...
...= \\\"\\\"\\npprof-addr = \\\"\\\"\\nstatus-addr = \\\"\\\"\\nclean-dump-file = true\\n\\n[from]\\n  host = \\\"127.0.0.1\\\"\\n  port = 3307\\n  user = \\\"root\\\"\\n  password = \\\"/Q7B9DizNLLTTfiZHv9WoEAKamfpIUs=\\\"\\n  max-allowed-packet = 67108864\\n\\n[to]\\n  host = \\\"127.0.0.1\\\"\\n  port = 4000\\n  user = \\\"root\\\"\\n  password = \\\"\\\"\\n  max-allowed-packet = 67108864\\n\\n[[route-rules]]\\n  schema-pattern = \\\"sharding*\\\"\\n  table-pattern = \\\"t*\\\"\\n  target-schema = \\\"db_target\\\"\\n  target-table = \\\"t_target\\\"\\n\\n[[route-rules]]\\n  schema-pattern = \\\"sharding*\\\"\\n  table-pattern = \\\"\\\"\\n  target-schema = \\\"db_target\\\"\\n  target-table = \\\"\\\"\\n\\n[[mapping-rule]]\\n  schema-pattern = \\\"sharding*\\\"\\n  table-pattern = \\\"t*\\\"\\n  source-column = \\\"id\\\"\\n  target-column = \\\"id\\\"\\n  expression = \\\"partition id\\\"\\n  arguments = [\\\"2\\\", \\\"sharding\\\", \\\"t\\\"]\\n  create-table-query = \\\"\\\"\\n\\n[block-allow-list]\\n  do-dbs = [\\\"~^sharding[\\\\\\\\d]+\\\"]\\n\\n  [[block-allow-list.do-tables]]\\n    db-name = \\\"~^sharding[\\\\\\\\d]+\\\"\\n    tbl-name = \\\"~^t[\\\\\\\\d]+\\\"\\n\"]}")
```
Comment (lance6716) [inline sync_diff_inspector/dm_test.go:164]: 😵how about using backquote string

**Review by lance6716 — COMMENTED** [review]
rest LGTM

File: sync_diff_inspector/diff.go (hunk 1 above)
Comment (WangXiangUSTC) [reply sync_diff_inspector/diff.go]: update in 6ade6b5

File: sync_diff_inspector/diff.go (hunk 3 above)
Comment (WangXiangUSTC) [reply sync_diff_inspector/diff.go]: support it in 6ade6b5

File: sync_diff_inspector/dm_test.go (hunk 4 above)
Comment (WangXiangUSTC) [reply sync_diff_inspector/dm_test.go:164]: just copy from dm's log, ignore it :cold_sweat:

File: sync_diff_inspector/diff.go (hunk 2 above)
Comment (WangXiangUSTC) [reply sync_diff_inspector/diff.go]: update in 6ade6b5

File: sync_diff_inspector/dm_test.go (hunk 4 above)
Comment (lance6716) [reply sync_diff_inspector/dm_test.go:164]: seems I don't have permission to resolve comments in this repo 🤔

File: sync_diff_inspector/diff.go [hunk 5]
```diff
@@ -100,6 +113,13 @@ func (df *Diff) init(cfg *Config) (err error) {
 
 // CreateDBConn creates db connections for source and target.
 func (df *Diff) CreateDBConn(cfg *Config) (err error) {
+	if len(df.subTaskCfgs) != 0 {
```
Comment (lance6716) [inline sync_diff_inspector/diff.go]: how about move this if block into `if len(cfg.DMAddr) != 0 {}` in `(*Diff).init`? so `CreateDBConn` won't do prepare jobs.

Comment (WangXiangUSTC) [reply sync_diff_inspector/diff.go]: update in fbe1068

File: sync_diff_inspector/dm_test.go (hunk 4 above)
Comment (WangXiangUSTC) [reply sync_diff_inspector/dm_test.go:164]: I resolve this conversation

**Comment by lance6716** [comment]
LGTM

I think till now all conversations could be resolved

File: sync_diff_inspector/config.go [hunk 6]
```diff
@@ -201,6 +202,11 @@ type Config struct {
 	// set true will continue check from the latest checkpoint
 	UseCheckpoint bool `toml:"use-checkpoint" json:"use-checkpoint"`
 
+	// DMAddr's format should like "http://127.0.0.1:8261"
```
Comment (csuzhangxc) [inline sync_diff_inspector/config.go]: Emphasize that this is the address of **DM-master**?

File: sync_diff_inspector/config.toml [hunk 7]
```diff
@@ -35,6 +35,10 @@ ignore-struct-check = false
 # the name of the file which saves sqls used to fix different data.
 fix-sql-file = "fix.sql"
 
+# check for dm's task
+#dm-addr = "http://127.0.0.1:8261"
+#dm-task = "test"
```
Comment (csuzhangxc) [inline sync_diff_inspector/config.toml]: How about adding more comments in this sample config file?

File: sync_diff_inspector/dm.go [hunk 8]
```diff
@@ -0,0 +1,76 @@
...
+}
+
+// getDMTaskCfg gets dm's sub task config
+func getDMTaskCfg(dmAddr, task string) ([]*config.SubTaskConfig, error) {
+	tr := &http.Transport{
+		// TODO: support tls
+		//TLSClientConfig: tlsCfg,
+	}
+	client := &http.Client{Transport: tr}
+	req, err := http.NewRequest("GET", getDMTaskCfgURL(dmAddr, task), nil)
```
Comment (csuzhangxc) [inline sync_diff_inspector/dm.go:46]: check this `err`?

File: sync_diff_inspector/dm.go [hunk 9]
```diff
@@ -0,0 +1,76 @@
...
+	body, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		return nil, err
+	}
+
+	getSubTaskCfgResp := &pb.GetSubTaskCfgResponse{}
+	err = json.Unmarshal(body, getSubTaskCfgResp)
+	if err != nil {
+		return nil, err
+	}
```
Comment (csuzhangxc) [inline sync_diff_inspector/dm.go:64]: Should we also check `getSubTaskCfgResp.Result`?

File: sync_diff_inspector/config.go [hunk 10]
```diff
@@ -283,40 +289,56 @@ func (c *Config) checkConfig() bool {
...
-	if c.TargetDBCfg.Snapshot != "" {
-		c.TargetDBCfg.Snapshot = strconv.Quote(c.TargetDBCfg.Snapshot)
-	}
-	if _, ok := sourceInstanceMap[c.TargetDBCfg.InstanceID]; ok {
-		log.Error("target has same instance id in source", zap.String("instance id", c.TargetDBCfg.InstanceID))
-		return false
-	}
+	// source DB, target DB and check table's information will get from DM, don't need to check them
+	if len(c.DMAddr) == 0 {
+		if len(c.SourceDBCfg) == 0 {
```
Comment (csuzhangxc) [inline sync_diff_inspector/config.go]: Should we need to log something if `len(c.DMAddr) != 0` and `len(c.SourceDBCfg) != 0` (or some check for `c.TargetDBCfg`)?

File: sync_diff_inspector/config.go (hunk 6 above)
Comment (WangXiangUSTC) [reply sync_diff_inspector/config.go]: update in https://github.com/pingcap/tidb-tools/pull/377/commits/e3bb4bf891d6ddb50b9f53e0b5e0f5b7184cc7f0

File: sync_diff_inspector/config.toml (hunk 7 above)
Comment (WangXiangUSTC) [reply sync_diff_inspector/config.toml]: I add a new example config file for DM https://github.com/pingcap/tidb-tools/pull/377/commits/e3bb4bf891d6ddb50b9f53e0b5e0f5b7184cc7f0

File: sync_diff_inspector/dm.go (hunk 9 above)
Comment (WangXiangUSTC) [reply sync_diff_inspector/dm.go:64]: add in https://github.com/pingcap/tidb-tools/pull/377/commits/e3bb4bf891d6ddb50b9f53e0b5e0f5b7184cc7f0

File: sync_diff_inspector/config.go (hunk 10 above)
Comment (WangXiangUSTC) [reply sync_diff_inspector/config.go]: add in https://github.com/pingcap/tidb-tools/pull/377/commits/e3bb4bf891d6ddb50b9f53e0b5e0f5b7184cc7f0

**Review by csuzhangxc — APPROVED** [review]
LGTM

**Pull request merged**